### PR TITLE
Fix Windows timer, mmap, and enable openmp

### PR DESCRIPTION
**Pull request by richinseattle**
This uses mman-win32, GetTickCount, and updates the OMP pragmas to be MSVC compatible with minimal lines changed in run.c favoring including the mman-win32.c/.h. The ugliest part is probably the loop iterators need to be declared before the OMP pragma. Added Makefile target for mingw32 cross compile and a .bat file for compiling with cl.exe from MSVC.

### CODE REVIEW

**Comment by karpathy** [comment]
@richinseattle , this PR now conflicts with @tairov PR here https://github.com/karpathy/llama2.c/pull/96 . I'm not an expert on this space, can you please describe the differences and why one might be preferred to the other? The one from @tairov looks a lot simpler.

File: Makefile [hunk 1]
```diff
@@ -32,6 +32,10 @@ runfast: run.c
 runomp: run.c
 	$(CC) -Ofast -fopenmp -march=native run.c  -lm  -o run
 
+.PHONT: win64
```
Comment (karpathy) [inline Makefile:35]: typo?

**Comment by richinseattle** [comment]
#96 reverts my original timer patch and adds mmap code from a GPL project. I mine uses the simplest 1-line timer change, adds the most commonly used mmap replacement for win32 which is MIT licensed, and I enabled openmp.

File: Makefile (hunk 1 above)
Comment (richinseattle) [reply Makefile]: oops, fixed.

File: build_msvc.bat [hunk 2]
```diff
@@ -0,0 +1 @@
+cl.exe /Ox /openmp /I. run.c mman-win32.c
```
Comment (karpathy) [inline build_msvc.bat:1]: Sorry noob question but how is this file related to Makefile? What is this file for / is it needed?

File: run.c [hunk 3]
```diff
@@ -190,8 +196,9 @@ void softmax(float* x, int size) {
 
 void matmul(float* xout, float* x, float* w, int n, int d) {
     // W (d,n) @ x (n,) -> xout (d,)
-    #pragma omp parallel for
-    for (int i = 0; i < d; i++) {
+    int i;
+    #pragma omp parallel for private(i)
```
Comment (karpathy) [inline run.c:200]: It doesn't work without the "for private(i)" part in Windows?

File: run.c [hunk 4]
```diff
@@ -361,13 +369,17 @@ int argmax(float* v, int n) {
 
 long time_in_ms() {
     struct timespec time;
+#ifndef _WIN32
```
Comment (karpathy) [inline run.c:372]: should this be moved up one row? as `time` is not used in the windows version?

Comment (richinseattle) [reply run.c:372]: Yea, originally I was using my previous timer definition on windows which used timespec struct, but changed to GetTickCount since that is standard pattern on windows, can move the ifdef

File: run.c (hunk 3 above)
Comment (richinseattle) [reply run.c:200]: correct, the counter variable has to be explicitly specified as private to each thread and declared before the pragma unfortunately.

File: build_msvc.bat (hunk 2 above)
Comment (richinseattle) [reply build_msvc.bat:1]: The Makefile is intended for cross compiling from linux, it will output a .exe. This .bat file would be what you run on windows if you have visual studio installed to generate an .exe from command line compiler cl.exe

File: run.c (hunk 4 above)
Comment (richinseattle) [reply run.c:372]: I can use the approach from #96 to overload the clock_gettime definition in a win.h if you want me to clean up the timer related ifdef

Comment (karpathy) [reply run.c:372]: Yeah, we could see what that looks like. I don't really understand how the overloading happens though. By including the file up top, the function down below gets overwritten?

File: Makefile [hunk 5]
```diff
@@ -32,6 +32,10 @@ runfast: run.c
 runomp: run.c
 	$(CC) -Ofast -fopenmp -march=native run.c  -lm  -o run
 
+.PHONY: win64
+win64: 
+	x86_64-w64-mingw32-gcc-win32 -Ofast -D_WIN32 -o run.exe -I. run.c mman-win32.c
```
Comment (karpathy) [inline Makefile:37]: is it common to use underscores like this? why not just -DWIN32

Comment (richinseattle) [reply Makefile:37]: _WIN32 is the compiler define, WIN32 is an application level define. I'm not sure if it is required with mingw tbh, it is defined automatically on MSVC.
see: https://stackoverflow.com/questions/662084/whats-the-difference-between-the-win32-and-win32-defines-in-c

**Comment by tairov** [comment]
Thanks @richinseattle for this PR. As it was suggested in the issues, this is the variation that uses [mman-win32](https://github.com/alitrack/mman-win32) lib. 
The little downside is that it requires 2 additions.
I would keep the `mmap-windows` approach for simplicity, if "public domain" license is not a blocker of course. 
As a side note, I would recommend to split PRs for compilation errors fix & openmp support for windows platform.
It's nice to see contributions from Windows/C experienced people 😃

File: Makefile (hunk 5 above)
Comment (karpathy) [reply Makefile:37]: I see thank you.

File: run.c (hunk 4 above)
Comment (richinseattle) [reply run.c:372]: I refactored and tested but it turns out that mingw doesn't have a timespec_get implementation so that would disallow cross-compilation. The best windows time_in_ms solution for time deltas is the ifdef I proposed here with GetTickCount()

Comment (karpathy) [reply run.c:372]: I'm not sure I understand. Isn't the timespec_get only used in the non-Windows version? So it should be moved inside the ifndef? It's spurious in Windows because we just call GetTickCount

Comment (richinseattle) [reply run.c:372]: On linux the function is clock_gettime(). My original patch proposed using timepsec_get because it exists on linux and MSVC and matches the same function prototype as clock_gettime(), and this was before I tried the cross-compile option which I think is convenient.

In theory the implementation could be hidden in a header that is only included on windows to implement clock_gettime() by using timespec_get() underneath. 

But that breaks mingw compiles, so we should stick with GetTickCount(), which does not return seconds+nano, so it needs to be placed in ifdefs in the time_in_ms() function since it returns ms resolution timing.

**Comment by richinseattle** [comment]
@tairov I wasn't sure what happens when pd code is included in GPL, but it looks like it is safe to cherry pick.
https://www.gnu.org/licenses/gpl-faq.en.html#CombinePublicDomainWithGPL

If @karpathy prefers your malloc that's fine, it only implements mmap and munmap, vs the complete implementation in the mman-win32. We could flatten all windows support code beyond a couple ifdefs into a single .c and .h regardless since it's MIT.

If we go that route, I'll pull your mmap changes and then resubmit with the openmp support, however the timer solution still needs to be GetTickCount() as I mentioned above because timespec_get() breaks on mingw compiles.

**Comment by tairov** [comment]
@richinseattle I totally agree and I got the point regarding `mingw & GetTickCount`. Isn't it worthwhile to create a separate PR for compatibility with `mingw` & `openmp`? It would be great to keep PRs atomic to the certain degree. As it should simplify reviews/reverts, etc.

**Comment by tairov** [comment]
@richinseattle also a qq: do you think `llama2.c` need "complete implementation in the mman-win32"? What are the downsides to have implementations for mmap and munmap only?

**Comment by richinseattle** [comment]
The complete API set is 
```
void*   mmap(void *addr, size_t len, int prot, int flags, int fildes, off_t off);
int     munmap(void *addr, size_t len);
int     mprotect(void *addr, size_t len, int prot);
int     msync(void *addr, size_t len, int flags);
int     mlock(const void *addr, size_t len);
int     munlock(const void *addr, size_t len);
```
I'm not strongly arguing you need the extra functions, just future proofing since it's not much extra code and standard practice for a mmap impl to have those APIs. There might be advantages to locking memory later (pretty sure they did that in llama.cpp).

**Comment by richinseattle** [comment]
Regardless of which mmap, I would propose we just add a win.c/.h or win-support.c/.h (whatever name you prefer) and stuff anything else that comes up in the future in there in addition to the memory routines.

**Comment by richinseattle** [comment]
Going to split this into 3 PRs

**Comment by richinseattle** [comment]
Okay, openmp and timer support have merged. I have sent #130 if you want to use the mman-win32 impl, or #96 can be revised. Only need to add the 2 lines to run.c and decide which mmap. #96 is a single .h file, mine splits the .h and .c anticipating future needs. I will send a final PR to update build scripts after this is sorted
